src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -843,13 +843,15 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<TypeSy
 
             checkNullability(containingSymbol, typeParameter, typeArgument, nullabilityDiagnosticsBuilderOpt);
 
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
             if (typeParameter.HasUnmanagedTypeConstraint)
             {
-                var managedKind = typeArgument.Type.ManagedKind;
+                var managedKind = typeArgument.Type.GetManagedKind(ref useSiteDiagnostics);
```

Comment:
>var managedKind = typeArgument.Type.GetManagedKind(ref useSiteDiagnostics); [](start = 16, length = 75)

I think it would be better to make changes only around this call. Report use-site diagnostics immediately and reset useSiteDiagnostics to null. Adding dedicated local would be preferred though. This will keep the change very localized and will significantly simplify a merge into features/UsedAssemblyReferences later.